gossip: maintain a gossip connectivity map

petermattis opened the pull request:
Maintain a map of gossip connectivity on each node and periodically log
this connectivity map.

Release note: None

petermattis committed:
gossip: propagate a connectivity map through gossip

Each node now periodically gossips the key `gossip-clients:<node-id>`
where the value is a comma separated list of nodes to which it has open
client connections. This key is gossipped at a fairly rapid frequency (2
secs) and with a low TTL (4 secs) so that nodes can have a very up to
date picture of the current connectivity of the gossip network. Note
that because it is only a single key being gossipped, the total data
volume this consumes will be very low in comparison to normal traffic or
to Raft hearbearts.

A future extension of this work would be for gossip to stop relying on
the sentinel key being gossipped and instead use a signal based on the
connectivity graph being fully connected.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/28963)
<!-- Reviewable:end -->

petermattis commented:
This extends the periodic gossip status logging to include a gossip connectivity map that can easily be converted into a dot graph. For example, on a 12-node local cluster this produces:

```
gossip connectivity
        n2 -> n1
        n3 -> n1
        n4 -> n1
        n5 -> n4
        n6 -> n4
        n7 -> n3
        n8 -> n3
        n9 -> n4
        n10 -> n2
        n11 -> n3
        n12 -> n5
```

Wrap that that connectivity stuff with `digraph D { ... }` and send it into dot to produce pretty graphs.

tschottdorf commented:
Haven't reviewed in detail though in principle I like this. Not crazy about putting this in the logs all the time? Perhaps add code that only prints this when the graph changes (can be as easy as comparing against the last string). You may also consider the format `n2 -> n1; n3 -> n1; ...` which I think is also valid dot.

petermattis commented:
Yeah, it does appear to work to put it all on one line. Also, this works towards adding a debug endpoint to graph gossip as well (you only need to touch one node instead of grabbing connectivity from all of those nodes). Mainly putting this out there to give a directional sense of some of my gossip thinking. With the gossip connectivity information each node could realize they have a disconnected gossip graph. I'll note that I think this PR as-is doesn't achieve this as the TTL for this info is too long. If a gossip partition were to occur, nodes would merrily think they are still connected.

petermattis commented:
Reworked the implementation so that it properly handles a node going down by increasing the frequency at which the gossip client info is gossipped and reducing the TTL. I've also updated the output to consume less vertical space per your suggestion:

```
gossip connectivity
  n1 [sentinel];
  n2 -> n1; n3 -> n1; n4 -> n1; n6 -> n3; n7 -> n4; n8 -> n3; n9 -> n4; n10 -> n3; n11 -> n2; n12 -> n10;
```

With just a small bit of editing this is compatible with dot:

```
digraph D {
  n1 [shape=doublecircle];
  n2 -> n1; n3 -> n1; n4 -> n1; n6 -> n3; n7 -> n4; n8 -> n3; n9 -> n4; n10 -> n3; n11 -> n2; n12 -> n10;
}
```

Which produces:

<img width="781" alt="screen shot 2018-08-22 at 4 56 35 pm" src="https://user-images.githubusercontent.com/6201335/44490460-68186080-a62c-11e8-8c54-a2b007a31680.png">

petermattis committed:
sql: add crdb_internal.gossip_network

Release note: None

petermattis committed:
roachtest: add gossip roachtest

The gossip roachtest starts a 9 node cluster and then randomly kills a
node and verifies that the gossip network stabilizes to the same state
on every node.

Release note: None

petermattis commented:
I've added two additional commits. The second adds a `crdb_internal.gossip_network` virtual table which exposes a nodes internal view of the gossip network. The third commit adds a roachtest which starts a 9 node cluster and then loops randomly killing a node and waiting for the gossip network on the remaining nodes to reach the same state. This seems to expose a bug in gossip:

```
> /Users/pmattis/Development/go/bin/roachprod stop local:3
local: stopping
1: gossip not ok (dead node 3 present): 2:1,3:1,4:1,5:3,6:4,8:5,9:5
1: gossip not ok (dead node 3 present): 2:1,3:1,4:1,5:3,6:4,7:1,8:5,9:5
1: gossip not ok (dead node 3 present): 2:1,3:1,4:1,5:3,6:4,7:1,8:5,9:5
1: gossip not ok (dead node 3 present): 2:1,3:1,4:1,6:4,7:1,8:5,9:5
5: gossip not ok: 2:1,4:1,6:4,7:1 != 8:5,9:5
5: gossip not ok: 2:1,4:1,6:4,7:1 != 8:5,9:5
5: gossip not ok: 2:1,4:1,6:4,7:1 != 8:5,9:5
5: gossip not ok: 2:1,4:1,6:4,7:1 != 8:5,9:5
```

I'm not sure if I'm more hopeful that there is something wrong with my test or the new gossip connectivity code, or if this actually found a way to reproduce gossip badness.

petermattis commented:
The partitions revealed by the roachtest are real, though they resolve themselves after a few minutes. Did the sentinel key need to timeout for the gossip partition to heal? I didn't time how long the condition lasted, but 2 min looks about right.

Also, I've seen CPU on one of the nodes go sky high while a partition is present. This happened only once, while I've seen partitions a half-dozen times now.

petermattis commented:
@spencerkimball or @a-robinson do either of you know why `sentinelGossipTTL` is set to 2 min? If a gossip partition occurs, that seems to be the duration for which the partition can continue to exist. And during the partition, liveness info won't be properly propagated throughout the cluster which seems bad.

a-robinson commented:
I can't say with certainty, but I'd guess that it was a reasonable setting when @spencerkimball first wrote the code and wasn't reconsidered when/since epoch-based range leases were added.

a-robinson commented:
Very nice find, by the way.

petermattis commented:
> I can't say with certainty, but I'd guess that it was a reasonable setting when @spencerkimball first wrote the code and wasn't reconsidered when/since epoch-based range leases were added.

Yeah, it looks like the 2m time was inherited from `clusterIDGossipTTL` in a42c4d7d503ad31b9fed29823d1ebf7ebb81bdbb (Jan 5, 2016). That setting changed from 30s to 2m in dbf6272fe6e706e9822e7fcbd1e031858d7c4c79 (May 12, 2015). All of this predates epoch-based leases (Oct 24, 2016).

PS The max time for a gossip partition is ~2m30s. 2m for the sentinel key to expire and 30s for the stall ticker to realize that the node is on the minority side of a partition and signal bootstrap to start making connections again. Even then, only a single connection is made per second (AFAICT).

spencerkimball commented:
The 2m setting is a function of that being a reasonable default from before epoch leases were a thing. We need to move that to be the same as epoch lease renewal. And sounds like we need to speed up the stall ticker so it’s on the same wavelength.

petermattis commented:
@spencerkimball Is there any danger to reducing the stall check interval? What I'm worried about is that when stalled we will wake the bootstrap loop too frequently which will try to make new connections over and over before we've had a chance to receive the sentinel key from the new connection. See the interaction between `Gossip.signalStallLocked` and `Gossip.bootstrap`.

petermattis commented:
FWIW, I'm testing with the sentinel gossip TTL reduced to `range-lease-duration / 2` (4.5s), the sentinel gossip interval set to `sentinel-TTL / 2` (2.25s), and the stall interval set to 2s. Everything seems to be happier in the new roachtest. Gossip partitions resolve themselves in a few seconds. The concern in the previous comment stems from an overabundance of caution.

petermattis committed:
gossip,storage: adjust gossip timings

Reduce the sentinel gossip TTL and interval from 2m / 1m, to 4.5s /
2.25s. The sentinel key informs gossip whether it is connected to the
primary gossip network or a partition and thus needs a short TTL so that
partitions are fixed quickly. In particular, partitions need to resolve
faster that the expiration-based lease timeout (9s) or node liveness
will be adversely affected.

Reduce the gossip stall check interval from 30s to 2s. The stall check
interval also affects how quickly gossip partitions are noticed and
repaired. The stall check itself is very cheap, so this produces no load
on the system.

Release note (bug fix): Reduce the duration of partitions in the gossip
network when a node crashes in order to eliminate a cause of temporary
data unavailability.

Fixes #27731

petermattis commented:
This is ready for a look.

petermattis commented:
@knz I'm looking for you to review the crdb_internal.gossip_network bits.

petermattis commented:
Added #29034 as the first commit here to remove test flakes in gossip under stress. I plan to merge that PR first.

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but look at my comment below. If you choose to go ahead and merge this as-is, please file a follow-up issue and assign to me.

Also you need to rebase to pick up changes to the show_trace logic test.

> Reviewed 3 of 3 files at r2, 6 of 6 files at r4, 2 of 2 files at r5, 4 of 4 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-:-LKkJGIE4ggrBbtz2wHo:b-tgsiez)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/crdb_internal.go, line 1986 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKkIC9A6s7KLKqPvNQf:-LKkIC9A6s7KLKqPvNQg:b-fpaq6j) ([raw file](https://github.com/cockroachdb/cockroach/blob/e91508c0b2b3fdcce07e369cfa07246543f175c1/pkg/sql/crdb_internal.go#L1986)):*
> ```Go
> var crdbInternalGossipNetworkTable = virtualSchemaTable{
> 	schema: `
> CREATE TABLE crdb_internal.gossip_network (
> ```

This table is reflecting the view of the gossip network from the perspective of the node where the table is queried, doesn't it?

If that is so, then we have a patter4n for that already:

- prefix the name of the table with `node_`
- include the node ID as first column `node_id`

(see e.g. `node_statement_statistics`, `node_build_info`, etc)

This way we can later extend the functionality to merge the contents of this table across all nodes to see a combined view *with the same columns*. 



(As I am writing this I realize the comment applies to the other gossip tables. Perhaps we should go back to those and change them in the same way. If you don't feel like doing so, I'll volunteer in a separate PR.)


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
TFTR

> Also you need to rebase to pick up changes to the show_trace logic test.

I think this PR is already rebased on latest master. The show_trace logic test has been flaky during the development of this PR. I've been meaning to have @andreimatei take a look.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-:-LKl4A6G6fJ5CzsXlPyR:b-4wrxwg)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/crdb_internal.go, line 1986 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKkIC9A6s7KLKqPvNQf:-LKl2r953CHCCleGEgBJ:b1nipy3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e91508c0b2b3fdcce07e369cfa07246543f175c1/pkg/sql/crdb_internal.go#L1986)):*
> This table is reflecting the view of the gossip network from the perspective of the node where the table is queried, doesn't it?

Yes.

> If that is so, then we have a pattern for that already:

As you note below, I was following the pattern of the other `crdb_internal.gossip_*` tables.

> This way we can later extend the functionality to merge the contents of this table across all nodes to see a combined view with the same columns.

For this table in particular, it is supposed to be identical across all nodes in the cluster (that is the point of the other commits in the PR). Any variation is temporary.

> (As I am writing this I realize the comment applies to the other gossip tables. Perhaps we should go back to those and change them in the same way. If you don't feel like doing so, I'll volunteer in a separate PR.)

The `gossip_` prefix implies being node specific to me, so I'm not sure if a `node_` prefix is necessary. Also, `node_gossip_nodes` would be a mouthful. Adding a `node_id` column seems reasonable, though outside of the scope of this PR. I'll file an issue if you like.


<!-- Sent from Reviewable.io -->

knz commented:
> I'll file an issue if you like.

Yes please.

Thanks otherwise for the explanation. It makes sense to me.
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

tbg reviewed (changes requested):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 2 of 3 files at r2, 12 of 12 files at r7, 6 of 6 files at r8, 2 of 2 files at r9, 4 of 4 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-:-LKvR-v43TBNGMyOpIuB:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/gossip/gossip.go, line 113 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKvNnJd4G7ZGYxSVEKC:-LKvNnJd4G7ZGYxSVEKD:b-x2ih3w) ([raw file](https://github.com/cockroachdb/cockroach/blob/29db75d2744fb643afdf27b611d92c7ccb9d6bb6/pkg/gossip/gossip.go#L113)):*
> ```Go
> 	defaultCullInterval = 60 * time.Second
> 
> 	// defaultClientsInterval ...
> ```

Assume this gets filled in in future revision.
___
*[pkg/gossip/gossip.go, line 588 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKvOKK_8h2cHoZg02XE:-LKvOKKaAKphaLsfVdZf:b-btjiex) ([raw file](https://github.com/cockroachdb/cockroach/blob/29db75d2744fb643afdf27b611d92c7ccb9d6bb6/pkg/gossip/gossip.go#L588)):*
> ```Go
> }
> 
> func (g *Gossip) connectivity() string {
> ```

Would you mind going the extra mile and giving this the same treatment as `g.clientStatus()`?
___
*[pkg/gossip/gossip.go, line 589 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKvQI8F0a1epWQfmEC9:-LKvQI8F0a1epWQfmECA:b1s7c8l) ([raw file](https://github.com/cockroachdb/cockroach/blob/34eb50a88e159230d97d4c6f379cef28183ad90f/pkg/gossip/gossip.go#L589)):*
> ```Go
> 
> func (g *Gossip) connectivity() string {
> 	ctx := g.AnnotateCtx(context.TODO())
> ```

Some of this code seems to overlap with the crdb_internal table. How about introducing a method that gives out the gossip connectivity in a predigested format?
___
*[pkg/gossip/keys.go, line 93 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKvO_GE8XRY9OusXPfr:-LKvO_GE8XRY9OusXPfs:bb7462z) ([raw file](https://github.com/cockroachdb/cockroach/blob/29db75d2744fb643afdf27b611d92c7ccb9d6bb6/pkg/gossip/keys.go#L93)):*
> ```Go
> 	KeyTableStatAddedPrefix = "table-stat-added"
> 
> 	// KeyGossipClientsPrefix ...
> ```

TODO
___
*[pkg/gossip/keys.go, line 137 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKvOa1o4TACYRXyFcEv:-LKvOa1o4TACYRXyFcEw:bb7462z) ([raw file](https://github.com/cockroachdb/cockroach/blob/29db75d2744fb643afdf27b611d92c7ccb9d6bb6/pkg/gossip/keys.go#L137)):*
> ```Go
> }
> 
> // MakeGossipClientsKey ...
> ```

TODO
___
*[pkg/sql/crdb_internal.go, line 1986 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKkIC9A6s7KLKqPvNQf:-LKvOdW054S6e5RiwFMs:brqyuwd) ([raw file](https://github.com/cockroachdb/cockroach/blob/e91508c0b2b3fdcce07e369cfa07246543f175c1/pkg/sql/crdb_internal.go#L1986)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

> This table is reflecting the view of the gossip network from the perspective of the node where the table is queried, doesn't it?

Yes.

> If that is so, then we have a pattern for that already:

As you note below, I was following the pattern of the other `crdb_internal.gossip_*` tables.

> This way we can later extend the functionality to merge the contents of this table across all nodes to see a combined view with the same columns.

For this table in particular, it is supposed to be identical across all nodes in the cluster (that is the point of the other commits in the PR). Any variation is temporary.

> (As I am writing this I realize the comment applies to the other gossip tables. Perhaps we should go back to those and change them in the same way. If you don't feel like doing so, I'll volunteer in a separate PR.)

The `gossip_` prefix implies being node specific to me, so I'm not sure if a `node_` prefix is necessary. Also, `node_gossip_nodes` would be a mouthful. Adding a `node_id` column seems reasonable, though outside of the scope of this PR. I'll file an issue if you like.
</blockquote></details>

I agree that this is a bit confusing, but I think the current situation is still reasonable. `gossip_` implies `node_` and I don't think it's worth renaming these tables as they're queried by the cli and there are compat constraints .
___
*[pkg/sql/crdb_internal.go, line 1983 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKvPRyo6O1h-VRGlfUh:-LKvPRyo6O1h-VRGlfUi:b277080) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec8214a6c469af87e4b0a265b19a1b296334561d/pkg/sql/crdb_internal.go#L1983)):*
> ```Go
> }
> 
> // crdbInternalGossipNetwork TODO(peter)
> ```

TODO

also add this to `debug zip` please.


<!-- Sent from Reviewable.io -->

tschottdorf commented:
PS we have gossip-connectivity related acceptance tests. Are you aware of those? You may want to either update them as well or remove them.

petermattis committed:
cli: add crdb_internal.gossip_network to "debug zip"

Release note: None

petermattis reviewed (commented):
> PS we have gossip-connectivity related acceptance tests. Are you aware of those? You may want to either update them as well or remove them.

Thanks for the reminder. I had forgotten about those acceptance tests. Definitely some overlap, but those tests didn't catch these gossip partition bugs which seem relatively easy to run into, so there worth is automatically suspect. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-:-LKvjCmR3q0FAGSo6I5H:b-ari48q)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/gossip/gossip.go, line 113 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKvNnJd4G7ZGYxSVEKC:-LKvb_QY4Gd6FODwqz9Y:b-qwgoyk) ([raw file](https://github.com/cockroachdb/cockroach/blob/29db75d2744fb643afdf27b611d92c7ccb9d6bb6/pkg/gossip/gossip.go#L113)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Assume this gets filled in in future revision.
</blockquote></details>

Heh, nope. An oversight. Fixed.
___
*[pkg/gossip/gossip.go, line 588 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKvOKK_8h2cHoZg02XE:-LKvga2x3kQxQtzp0Hvw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/29db75d2744fb643afdf27b611d92c7ccb9d6bb6/pkg/gossip/gossip.go#L588)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Would you mind going the extra mile and giving this the same treatment as `g.clientStatus()`?
</blockquote></details>

Done.
___
*[pkg/gossip/gossip.go, line 589 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKvQI8F0a1epWQfmEC9:-LKvgalr44wGa-Ao86iF:b6n9wlc) ([raw file](https://github.com/cockroachdb/cockroach/blob/34eb50a88e159230d97d4c6f379cef28183ad90f/pkg/gossip/gossip.go#L589)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Some of this code seems to overlap with the crdb_internal table. How about introducing a method that gives out the gossip connectivity in a predigested format?
</blockquote></details>

`Gossip.Connectivity` now does it exactly that.
___
*[pkg/gossip/keys.go, line 93 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKvO_GE8XRY9OusXPfr:-LKvhrzk0GEIoeE67dAf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/29db75d2744fb643afdf27b611d92c7ccb9d6bb6/pkg/gossip/keys.go#L93)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

TODO
</blockquote></details>

Done.
___
*[pkg/gossip/keys.go, line 137 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKvOa1o4TACYRXyFcEv:-LKvhzfL0Q8X7yM5On4N:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/29db75d2744fb643afdf27b611d92c7ccb9d6bb6/pkg/gossip/keys.go#L137)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

TODO
</blockquote></details>

Done.
___
*[pkg/sql/crdb_internal.go, line 1983 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKvPRyo6O1h-VRGlfUh:-LKvirvl0rzFDo3b7koy:bjk3q99) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec8214a6c469af87e4b0a265b19a1b296334561d/pkg/sql/crdb_internal.go#L1983)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

TODO

also add this to `debug zip` please.
</blockquote></details>

Fixed and added to `debug zip`.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> Definitely some overlap, but those tests didn't catch these gossip partition bugs which seem relatively easy to run into, so there worth is automatically suspect.

The reason those tests weren't catching gossip partitions is because they check for the `node:` keys being gossiped, but those keys have a TTL of 2h. So as long as you can connect to a node that had those keys at some point, the test will succeed even if a partition is now present. I'm going to file an issue to migrate those acceptance tests to roachtest and expand on the checks they perform. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-:-LKvjf4AAriRPp4VJKKw:b88ellm)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
https://github.com/cockroachdb/cockroach/issues/29115

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-:-LKvlWL-0zfoscBQLgOj:b-l78024)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

a-robinson reviewed (changes requested):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

This is definitely feeling more fragile than I thought it was a week or two ago. Relying on the sentinel gossip making it from its source to all other nodes in 2.25 seconds in order to avoid gossip thinking it's orphaned is pretty strict. And even with Nathan's recent change to improve the locking with a `RWLock`, there are still a lot of things that grab `Gossip.mu` (t's what got most backed up in my tpc-c 10k cluster with vmodule on, albeit without Nathan's change). I look forward to this getting some test cycles in under heavy client/network load and on geographically distributed clusters.

> Reviewed 1 of 1 files at r1, 1 of 3 files at r2, 16 of 16 files at r11, 6 of 6 files at r12, 2 of 2 files at r13, 4 of 4 files at r14, 2 of 2 files at r15.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-:-LKw0uxsB0FkNOz2s-HZ:b56633x)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale)

___
*[pkg/gossip/gossip.go, line 612 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKvygWU-XaD9EkIXJqH:-LKvygWU-XaD9EkIXJqI:b-fzcgd1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bb96e280e09d8e18fd7097260cea26ed773a09b/pkg/gossip/gossip.go#L612)):*
> ```Go
> 		v, err := i.Value.GetBytes()
> 		if err != nil {
> 			log.Errorf(ctx, "unable to retrieve gossip value for %s: %v",
> ```

As far as I can tell, we only remove nodes from `g.nodeDescs` when a new node comes up at the address of the old node. Doesn't that mean that if a node goes offline for a while that we'll still have a `g.nodeDescs` entry for it but won't have a gossip clients key for it (due to the TTL expiring)? This log statement could get very repetitive if I'm not misunderstanding.
___
*[pkg/gossip/gossip.go, line 925 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKvzlWo8TX9euvbAg9v:-LKvzlWp-EvyXv5wkzgE:bgzsfzn) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bb96e280e09d8e18fd7097260cea26ed773a09b/pkg/gossip/gossip.go#L925)):*
> ```Go
> 	for _, c := range g.clientsMu.clients {
> 		if c.peerID != 0 {
> 			fmt.Fprintf(&buf, "%s%d", sep, c.peerID)
> ```

I assume we're using a hand-crafted string rather than a proto for readability when dumping the raw gossip values? Otherwise it doesn't make much sense.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
We gossip the sentinel every 2.25 secs and time it out after 4.5s. Perhaps those numbers should be adjusted, but they need to be set in relation to the node liveness timeout of 9s. The previous sentinel gossip interval and TTL did not make any sense as they detected gossip partitions minutes after they occurred. I'm open to other arguments for different settings here. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-:-LKw5q6T9WSq_eLJWWow:bl4myjl)* status: :shipit: complete! 2 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/gossip/gossip.go, line 612 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKvygWU-XaD9EkIXJqH:-LKw5J7b0mBqj_t0M-RF:bnu8r1f) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bb96e280e09d8e18fd7097260cea26ed773a09b/pkg/gossip/gossip.go#L612)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

As far as I can tell, we only remove nodes from `g.nodeDescs` when a new node comes up at the address of the old node. Doesn't that mean that if a node goes offline for a while that we'll still have a `g.nodeDescs` entry for it but won't have a gossip clients key for it (due to the TTL expiring)? This log statement could get very repetitive if I'm not misunderstanding.
</blockquote></details>

This is the `gossip-clients` key which expires quickly and should only ever contain bytes data. This log message should never occur.
___
*[pkg/gossip/gossip.go, line 925 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKvzlWo8TX9euvbAg9v:-LKw5TJK5WMydckET8cF:be4gtwy) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bb96e280e09d8e18fd7097260cea26ed773a09b/pkg/gossip/gossip.go#L925)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I assume we're using a hand-crafted string rather than a proto for readability when dumping the raw gossip values? Otherwise it doesn't make much sense.
</blockquote></details>

Well, it started out for readability, but then the PR evolved substantially. I suppose I could change it to a proto, but that feels like overkill here. 


<!-- Sent from Reviewable.io -->

a-robinson reviewed (approved):
Yeah, I'm not saying that it is more fragile now (it's certainly safer now than before), just that its fragility is now much more apparent.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-:-LKw7sWc3CjrXcKAqTrm:b-5eyyyn)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale)

___
*[pkg/gossip/gossip.go, line 612 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKvygWU-XaD9EkIXJqH:-LKw8Tfh2wogAwlAthtm:bjcodo3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bb96e280e09d8e18fd7097260cea26ed773a09b/pkg/gossip/gossip.go#L612)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This is the `gossip-clients` key which expires quickly and should only ever contain bytes data. This log message should never occur.
</blockquote></details>

Whoops, I thought this was being logged for the `i == nil` case. Never mind.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> Yeah, I'm not saying that it is more fragile now (it's certainly safer now than before), just that its fragility is now much more apparent.

Ack. We'll definitely want to pay close attention to gossip-related badness over the next few weeks. I also want to do more gossip-specific testing. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-:-LKwCStt9GVjhlwRWlwh:bken4mr)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale)

___
*[pkg/gossip/gossip.go, line 612 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKvygWU-XaD9EkIXJqH:-LKwCghj36oHZznrqguT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bb96e280e09d8e18fd7097260cea26ed773a09b/pkg/gossip/gossip.go#L612)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Whoops, I thought this was being logged for the `i == nil` case. Never mind.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
Still LGTM but I did have a few comments I think are worth looking at before merge.

> Reviewed 4 of 16 files at r11, 14 of 14 files at r16, 6 of 6 files at r17, 2 of 2 files at r18, 4 of 4 files at r19, 2 of 2 files at r20.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-:-LKw_fWDBID2o1B7QEPY:b-beryyr)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale)

___
*[pkg/gossip/gossip.go, line 604 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKwZ4_867_bImECn0VY:-LKwZ4_867_bImECn0VZ:bsfvoyj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f928811b7ce34d34c0239d95396c3782ead2c173/pkg/gossip/gossip.go#L604)):*
> ```Go
> 	}
> 
> 	for nodeID := range g.nodeDescs {
> ```

Why not use `func (g *Gossip) IterateInfos(prefix string, visit func(k string, info Info) error) error {` instead of discovering NodeIDs first and then looking at entries?
___
*[pkg/gossip/gossip.proto, line 117 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKw_ARnCtC4Qtq55fbu:-LKw_ARnCtC4Qtq55fbv:br2kwv8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f928811b7ce34d34c0239d95396c3782ead2c173/pkg/gossip/gossip.proto#L117)):*
> ```protobuf
> }
> 
> message Connectivity {
> ```

Stick this into GetInfoStatus() while you're here.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28963)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale)

___
*[pkg/gossip/gossip.go, line 604 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKwZ4_867_bImECn0VY:-LKwaZgU68izBbWuS86Y:bdq0ssb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f928811b7ce34d34c0239d95396c3782ead2c173/pkg/gossip/gossip.go#L604)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Why not use `func (g *Gossip) IterateInfos(prefix string, visit func(k string, info Info) error) error {` instead of discovering NodeIDs first and then looking at entries?
</blockquote></details>

`IterateInfos` loops over all of the infos, while this only looks at infos that are possibly of interest. Probably doesn't matter. Give me one more push and I'll change to use `IterateInfos`. 
___
*[pkg/gossip/gossip.proto, line 117 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKw_ARnCtC4Qtq55fbu:-LKwaEwsDNtlvz9kHF3V:bqf0pzj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f928811b7ce34d34c0239d95396c3782ead2c173/pkg/gossip/gossip.proto#L117)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Stick this into GetInfoStatus() while you're here.
</blockquote></details>

I stuck it in `InfoStatus`, but forgot to populate. Done.


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
> Reviewed 14 of 14 files at r21, 6 of 6 files at r22, 2 of 2 files at r23, 4 of 4 files at r24, 2 of 2 files at r25.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28963)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale)

___
*[pkg/gossip/gossip.go, line 604 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKwZ4_867_bImECn0VY:-LKwbGPYBwHgieafbDak:bmziohn) ([raw file](https://github.com/cockroachdb/cockroach/blob/f928811b7ce34d34c0239d95396c3782ead2c173/pkg/gossip/gossip.go#L604)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`IterateInfos` loops over all of the infos, while this only looks at infos that are possibly of interest. Probably doesn't matter. Give me one more push and I'll change to use `IterateInfos`. 
</blockquote></details>

No, fine by me. There's no way something could be gossiped and we don't have a nodeDesc, right?


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28963)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale)

___
*[pkg/gossip/gossip.go, line 604 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKwZ4_867_bImECn0VY:-LKwcSVG6AEeUm4ALN0p:b7mfhq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f928811b7ce34d34c0239d95396c3782ead2c173/pkg/gossip/gossip.go#L604)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

No, fine by me. There's no way something could be gossiped and we don't have a nodeDesc, right?
</blockquote></details>

We gossip the node descriptor as the first thing right at system startup and then every hour afterwards. So no, we shouldn't have a `gossip-client` key without the associated node descriptor. 


<!-- Sent from Reviewable.io -->

petermattis commented:
I'm baffled by the `show_trace` logic test failure. I've seen it locally, but been unable to reliably reproduce it. It is not at all clear how it is related to this PR.

petermattis commented:
Ah-ha! Here is a magical incantation to reproduce the failure quickly: `make stress PKG=./sql/logictest TESTS='TestLogic/^(local|local-opt)$$/^show_trace$$' TESTFLAGS="-v" STRESSFLAGS="-p 24"`

petermattis commented:
The flakiness is caused by reducing the sentinel gossip interval from 2m to 4.5s. That causes `Replica.maybeGossipFirstRange` to be called every 4.5s and, on the leaseholder, for `Replica.gossipFirstRange` to be called every 4.5s.

Gossiping the sentinel is fine, but `gossipFirstRange` also gossips the range descriptor for the first range. `DistSender` has registered a callback so that it evicts the range descriptor for the first range whenever that gossip key changes.

@nvanbenschoten Rather than always evicting the first range descriptor upon a gossip callback, could we only evict it if it has changed?

nvanbenschoten commented:
> Rather than always evicting the first range descriptor upon a gossip callback, could we only evict it if it has changed?

I don't see any problem with that. It's probably a good change regardless of whether it fixes flakiness or not.

petermattis committed:
gossip: do not invoke callbacks if an info's value is unchanged

Previously, gossip would invoke a callback whenever a new info that
matched the callback pattern was received, even if the info's value was
unchanged. There were a handful of gossip uses (the table-stats cache
and the store-pool) that were relying on this behavior which now opt-in
to it via the `gossip.Redundant` callback option.

Release note: None

petermattis commented:
> I don't see any problem with that. It's probably a good change regardless of whether it fixes flakiness or not.

This does fix the flakiness.

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28963)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale)

___
*[pkg/kv/dist_sender.go, line 264 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKwvQrjEmh8TELylXs_:-LKwvQrl8DpWuKR_yW2j:b-9xyj9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa626efce649786fca82c932e7e4481cec63a640/pkg/kv/dist_sender.go#L264)):*
> ```Go
> 					mu.value = append([]byte(nil), value.RawBytes...)
> 				}
> 				mu.Unlock()
> ```

This is almost too hacky for me. I'm investigating pushing this into gossip. I don't think there is any need for gossip to invoke a callback if the value hasn't changed. Is anything relying on gossip callbacks in a timely manner for unchanged data. I'll find out soon.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28963)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale)

___
*[pkg/kv/dist_sender.go, line 264 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LKwvQrjEmh8TELylXs_:-LKxr-QECIwmyYd2HiVB:b-rs5u41) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa626efce649786fca82c932e7e4481cec63a640/pkg/kv/dist_sender.go#L264)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This is almost too hacky for me. I'm investigating pushing this into gossip. I don't think there is any need for gossip to invoke a callback if the value hasn't changed. Is anything relying on gossip callbacks in a timely manner for unchanged data. I'll find out soon.
</blockquote></details>

Ok, the last 2 commits take different approaches to solving the over-aggressive eviction of the first range descriptor from the range descriptor cache. The last commit "de-bounces" all gossip callbacks so that they are only invoked when the value changes. This required adding a bypass mechanism for various tests which expect the callback even when nothing has changed. The other commit more narrowly targets the range cache eviction in `DistSender`. Thoughts?


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I assume you've thoroughly checked all tests that you stuck the legacy behavior in to make sure that we're not masking anything that would break in real code.

> Reviewed 1 of 1 files at r26, 11 of 11 files at r27.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-:-LKzifuw-qvNgDRjKeyz:belbkh1)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale)



<!-- Sent from Reviewable.io -->

petermattis commented:
>  I assume you've thoroughly checked all tests that you stuck the legacy behavior in to make sure that we're not masking anything that would break in real code.

I originally started trying to fix all of the tests individually, but most of them have similar patterns of code of registering a gossip callback and then waiting for the callback to fire. This isn't universally true, though. I need to double-check `TestStoreRangeRemoveDead` and `TestStoreSetRangesMaxBytes`.

The CI `TestDistSQLRangeCachesIntegrationTest` flake is real and reproducible. I need to track that down. It's unrelated to the gossip callback de-bouncing.

petermattis commented:
Changed `TestStoreSetRangesMaxBytes` to gossip a non-empty `SystemConfig` to workaround the de-bouncing for the test.

`TestStoreRangeRemoveDead` is more interesting. The test is verifying that a store that is not gossiped regularly will be marked as dead. The test accomplishes by only gossiping the store descriptors for 3 of the 4 stores in the test cluster, but the 3 descriptors it is gossiping are unchanging so the store-pool is seeing all 4 stores as dead. I've replaced `Gossip.EnableRedundantCallbacks` with an option mechanism to `Gossip.RegisterCallback` so that an individual registration can not opt into redundant callbacks. PTAL.

Changed `TableStatisticsCache` to opt-in to redundant callbacks as it is using them as a signal that new stats for a table have been added.

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28963)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale)

___
*[pkg/sql/distsql_physical_planner_test.go, line 314 at r33](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LL0f1tg-uBDBGsZ0IKx:-LL0f1tg-uBDBGsZ0IKy:b-gmev8j) ([raw file](https://github.com/cockroachdb/cockroach/blob/557e690062cce08052912d842e79af39c85105de/pkg/sql/distsql_physical_planner_test.go#L314)):*
> ```Go
> 	// the unpredictable nature of those updates interferes with the expectations
> 	// of this test below.
> 	tc.Server(3).DistSender().DisableFirstRangeUpdates()
> ```

@andrei Here is how I'm cutting the wire.


<!-- Sent from Reviewable.io -->

petermattis commented:
Apologies to the reviewers that this PR has ballooned into a behemoth. I think it is good to go now. Can someone give it another check to see if I missed anything or if there are other concerns?

knz reviewed (commented):
sql and storage changes <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> - I do not know how to review the rest.

> Reviewed 6 of 6 files at r29, 3 of 4 files at r31, 2 of 2 files at r32, 6 of 12 files at r33, 1 of 1 files at r34.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-:-LL1jn6ZE7sAfGxqJvB_:bouvpl5)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale)



<!-- Sent from Reviewable.io -->

petermattis commented:
@tschottdorf or @a-robinson did you want to take a final look at this PR?

tbg reviewed (commented):
> Reviewed 24 of 24 files at r28, 6 of 6 files at r29, 2 of 2 files at r30, 4 of 4 files at r31, 2 of 2 files at r32, 11 of 12 files at r33, 1 of 1 files at r34.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28963)* status: :shipit: complete! 2 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/gossip/infostore_test.go, line 443 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LL5cLH41JKWgwMJ294O:-LL5cLH41JKWgwMJ294P:b-hni9a) ([raw file](https://github.com/cockroachdb/cockroach/blob/c89d5f71d9d1dede22bb6d29ab7b5af27e712080/pkg/gossip/infostore_test.go#L443)):*
> ```Go
> 	}
> 
> 	// Update an info twice. The first time we expect to see a callback. The
> ```

No coverage for registrations with the redundant flag?
___
*[pkg/kv/dist_sender_test.go, line 814 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LL5cCT469yqp_B1iHYh:-LL5cCT469yqp_B1iHYi:bx2uain) ([raw file](https://github.com/cockroachdb/cockroach/blob/c89d5f71d9d1dede22bb6d29ab7b5af27e712080/pkg/kv/dist_sender_test.go#L814)):*
> ```Go
> 		}
> 	}
> 	// Tweak the descriptor so that the gossip callback will be invoked.
> ```

Necessary now?
___
*[pkg/storage/allocator_test.go, line 5304 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LL5c2RcEh1UQWbkWoD5:-LL5c2RcEh1UQWbkWoD6:b7r5xui) ([raw file](https://github.com/cockroachdb/cockroach/blob/c89d5f71d9d1dede22bb6d29ab7b5af27e712080/pkg/storage/allocator_test.go#L5304)):*
> ```Go
> 	g.RegisterCallback(gossip.MakePrefixPattern(gossip.KeyStorePrefix),
> 		func(_ string, _ roachpb.Value) { wg.Done() },
> 		gossip.Redundant)
> ```

Comment
___
*[pkg/storage/allocator_test.go, line 5442 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LL5c35b-fdmLIuqT3q8:-LL5c35b-fdmLIuqT3q9:b7r5xui) ([raw file](https://github.com/cockroachdb/cockroach/blob/c89d5f71d9d1dede22bb6d29ab7b5af27e712080/pkg/storage/allocator_test.go#L5442)):*
> ```Go
> 	g.RegisterCallback(gossip.MakePrefixPattern(gossip.KeyStorePrefix),
> 		func(_ string, _ roachpb.Value) { wg.Done() },
> 		gossip.Redundant)
> ```

Comment
___
*[pkg/storage/gossip_test.go, line 55 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LL5c54o1vFtJhBOeLsx:-LL5c54o1vFtJhBOeLsy:b7r5xui) ([raw file](https://github.com/cockroachdb/cockroach/blob/c89d5f71d9d1dede22bb6d29ab7b5af27e712080/pkg/storage/gossip_test.go#L55)):*
> ```Go
> 			}
> 		},
> 		gossip.Redundant,
> ```

Comment
___
*[pkg/testutils/gossiputil/store_gossiper.go, line 49 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LL5c7Rw6pN2-koxdKM_:-LL5c7Rw6pN2-koxdKMa:b7r5xui) ([raw file](https://github.com/cockroachdb/cockroach/blob/c89d5f71d9d1dede22bb6d29ab7b5af27e712080/pkg/testutils/gossiputil/store_gossiper.go#L49)):*
> ```Go
> 		delete(sg.storeKeyMap, key)
> 		sg.cond.Broadcast()
> 	}, gossip.Redundant)
> ```

Comment


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-:-LL5kkicARdeiucCwIpy:b-814g8k)* status: :shipit: complete! 2 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/gossip/infostore_test.go, line 443 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LL5cLH41JKWgwMJ294O:-LL5kbi24c846DRjGAyl:b-d8b6g9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c89d5f71d9d1dede22bb6d29ab7b5af27e712080/pkg/gossip/infostore_test.go#L443)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

No coverage for registrations with the redundant flag?
</blockquote></details>

Done. I've made `cbAll` take the `Redundant` option which required tweaking the logic in the loop below slightly.
___
*[pkg/kv/dist_sender_test.go, line 814 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LL5cCT469yqp_B1iHYh:-LL5d78cE-RLhcMjQYib:b-vyeh2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/c89d5f71d9d1dede22bb6d29ab7b5af27e712080/pkg/kv/dist_sender_test.go#L814)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Necessary now?
</blockquote></details>

Yes, because we de-bounce the callbacks for `KeyFirstRangeDescriptor` in `DistSender`.
___
*[pkg/storage/allocator_test.go, line 5304 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LL5c2RcEh1UQWbkWoD5:-LL5jFp68gAScPPy5Akb:blcqx3j) ([raw file](https://github.com/cockroachdb/cockroach/blob/c89d5f71d9d1dede22bb6d29ab7b5af27e712080/pkg/storage/allocator_test.go#L5304)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Comment
</blockquote></details>

Done. I thought the `wg.Done()` on the previous line made that clear. My comment isn't very satisfactory. Let me know if you have a way to make it clearer.
___
*[pkg/storage/allocator_test.go, line 5442 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LL5c35b-fdmLIuqT3q8:-LL5jVrCBBLINAKMRII-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c89d5f71d9d1dede22bb6d29ab7b5af27e712080/pkg/storage/allocator_test.go#L5442)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Comment
</blockquote></details>

Done.
___
*[pkg/storage/gossip_test.go, line 55 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LL5c54o1vFtJhBOeLsx:-LL5jdJ76VmlYYZ1091N:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c89d5f71d9d1dede22bb6d29ab7b5af27e712080/pkg/storage/gossip_test.go#L55)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Comment
</blockquote></details>

Done.
___
*[pkg/testutils/gossiputil/store_gossiper.go, line 49 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/28963#-LL5c7Rw6pN2-koxdKM_:-LL5jqid6Bz94L7DdfDa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c89d5f71d9d1dede22bb6d29ab7b5af27e712080/pkg/testutils/gossiputil/store_gossiper.go#L49)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Comment
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

petermattis commented:
🤞 

bors r=knz,tschottdorf,a-robinson

petermattis commented:
Fixed stupid flake I added to `gossip.TestCallbacks`. Trying again.

bors r=knz,tschottdorf,a-robinson
